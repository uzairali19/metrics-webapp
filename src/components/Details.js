import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import {
  ImageList,
  ImageListItem,
  Typography,
  AppBar,
  Toolbar,
  AccordionDetails,
  AccordionSummary,
  Accordion,
} from '@mui/material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';

const Details = () => {
  const [expanded, setExpanded] = React.useState(false);
  const { id } = useParams();
  const marvels = useSelector((state) => state.marvelsReducer);
  const { results } = marvels.data;
  const data = results.find((v) => v.id === Number(id));

  const handleChange = (panel) => (event, isExpanded) => {
    setExpanded(isExpanded ? panel : false);
  };

  return (
    <div>
      {data.id === Number(id) ? (
        <div>
          <ImageList sx={{ width: 'auto', gap: '0 !important' }} variant="quilted" cols={2}>
            <ImageListItem cols={2} rows={2}>
              <img
                src={`${data.thumbnail.path}.${data.thumbnail.extension}`}
                alt={data.name}
                loading="lazy"
                style={{ width: '100%' }}
              />
            </ImageListItem>
          </ImageList>
          <div className="detail-title">{data.name}</div>
          <div className="detail-description">
            {data.description ? data.description : 'Description not avaiable for this character'}
          </div>
          <div className="detail-comics">
            <Accordion expanded={expanded === 'panel1'} onChange={handleChange('panel1')}>
              <AccordionSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1bh-content"
                id="panel1bh-header"
              >
                <Typography sx={{ width: '33%', flexShrink: 0 }}>Comics</Typography>
                <Typography sx={{ color: 'text.secondary' }}>
                  Total Comics:
                  {data.comics.available}
                </Typography>
              </AccordionSummary>
              {data.comics.items.map((v) => {
                return (
                  <div key={v.name}>
                    <AccordionDetails>
                      <Typography>{v.name}</Typography>
                    </AccordionDetails>
                  </div>
                );
              })}
            </Accordion>
          </div>
          <div className="detail-events">
            <Accordion expanded={expanded === 'panel2'} onChange={handleChange('panel2')}>
              <AccordionSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1bh-content"
                id="panel1bh-header"
              >
                <Typography sx={{ width: '33%', flexShrink: 0 }}>Events</Typography>
                <Typography sx={{ color: 'text.secondary' }}>
                  Total Events:
                  {data.events.available}
                </Typography>
              </AccordionSummary>
              {data.events.items.map((v) => {
                return (
                  <div key={v.name}>
                    <AccordionDetails>
                      <Typography>{v.name}</Typography>
                    </AccordionDetails>
                  </div>
                );
              })}
            </Accordion>
          </div>
          <div className="detail-stories">
            <Accordion expanded={expanded === 'panel3'} onChange={handleChange('panel3')}>
              <AccordionSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1bh-content"
                id="panel1bh-header"
              >
                <Typography sx={{ width: '33%', flexShrink: 0 }}>Stories</Typography>
                <Typography sx={{ color: 'text.secondary' }}>
                  Total Stories:
                  {data.stories.available}
                </Typography>
              </AccordionSummary>
              {data.stories.items.map((v) => {
                return (
                  <div key={v.name}>
                    <AccordionDetails>
                      <Typography>{v.name}</Typography>
                    </AccordionDetails>
                  </div>
                );
              })}
            </Accordion>
          </div>
          <div className="detail-series">
            <Accordion expanded={expanded === 'panel4'} onChange={handleChange('panel4')}>
              <AccordionSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1bh-content"
                id="panel1bh-header"
              >
                <Typography sx={{ width: '33%', flexShrink: 0 }}>Series</Typography>
                <Typography sx={{ color: 'text.secondary' }}>
                  Total Series:
                  {data.comics.available}
                </Typography>
              </AccordionSummary>
              {data.series.items.map((v) => {
                return (
                  <div key={v.name}>
                    <AccordionDetails>
                      <Typography>{v.name}</Typography>
                    </AccordionDetails>
                  </div>
                );
              })}
            </Accordion>
          </div>
          <Typography variant="h6" className="more-detail">
            Find more details about this character at:
            <ul>
              {data.urls.map((v) => {
                return (
                  <li key={v.type}>
                    <a href={v.url} target="_blank" rel="noopener noreferrer">
                      {v.type}
                    </a>
                  </li>
                );
              })}
            </ul>
          </Typography>
          <AppBar position="fixed" sx={{ top: 'auto', bottom: 0, backgroundColor: '#000' }}>
            <Toolbar>
              <Typography variant="h6" color="inherit">
                {marvels.attributionText}
              </Typography>
            </Toolbar>
          </AppBar>
        </div>
      ) : null}
    </div>
  );
};

export default Details;
