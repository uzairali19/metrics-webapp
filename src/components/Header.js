import * as React from 'react';
import { styled, alpha } from '@mui/material/styles';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import InputBase from '@mui/material/InputBase';
import SearchIcon from '@mui/icons-material/Search';
import ArrowBackIosNewIcon from '@mui/icons-material/ArrowBackIosNew';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const Search = styled('div')(({ theme }) => ({
  position: 'relative',
  borderRadius: theme.shape.borderRadius,
  backgroundColor: alpha(theme.palette.common.white, 0.15),
  '&:hover': {
    backgroundColor: alpha(theme.palette.common.white, 0.25),
  },
  marginLeft: '0',
  width: '100%',
}));

const SearchIconWrapper = styled('div')(({ theme }) => ({
  padding: theme.spacing(0, 2),
  height: '100%',
  position: 'absolute',
  pointerEvents: 'none',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
}));

const StyledInputBase = styled(InputBase)(({ theme }) => ({
  color: 'inherit',
  '& .MuiInputBase-input': {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `calc(1em + ${theme.spacing(4)})`,
    transition: theme.transitions.create('width'),
    width: '100%',
  },
}));

const Header = ({ set }) => (
  <Box sx={{ flexGrow: 1 }}>
    <AppBar sx={{ backgroundColor: '#121212' }} position="static">
      <Toolbar>
        <IconButton size="large" edge="start" aria-label="open drawer">
          <Link className="arrow-link" to="/">
            <ArrowBackIosNewIcon sx={{ color: '#fff' }} />
          </Link>
        </IconButton>
        <Typography
          variant="h6"
          wrap
          component="div"
          style={{ width: '100%' }}
          sx={{ flexGrow: 1, display: { sm: 'block' } }}
        >
          Heros App
        </Typography>
        <Search>
          <SearchIconWrapper>
            <SearchIcon />
          </SearchIconWrapper>
          <StyledInputBase
            placeholder="Searchâ€¦"
            inputProps={{ 'aria-label': 'search' }}
            onChange={(e) => set(e.target.value)}
          />
        </Search>
      </Toolbar>
    </AppBar>
  </Box>
);

Header.propTypes = {
  set: PropTypes.func,
};

Header.defaultProps = {
  set: () => {},
};

export default Header;
