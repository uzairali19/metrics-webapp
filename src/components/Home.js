import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import { ImageList, ImageListItem, ImageListItemBar } from '@mui/material';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const Home = ({ search }) => {
  const [loading, setLoading] = useState(false);
  const [data, setData] = useState([]);
  const marvels = useSelector((state) => state.marvelsReducer);

  useEffect(() => {
    const data = marvels.data.results;
    setLoading(true);
    setData(data);
  }, [marvels]);

  return (
    <div>
      {search === '' ? (
        <div>
          <ImageList sx={{ width: 'auto', gap: '0 !important' }} variant="quilted" cols={2}>
            {loading ? (
              <ImageListItem sx={{ overflow: 'hidden' }} cols={2} rows={2}>
                <Link to={`/${data[39].id}`}>
                  <img
                    src={`${data[39].thumbnail.path}.${data[39].thumbnail.extension}`}
                    alt={data[39].name}
                    loading="lazy"
                    style={{ width: '100%' }}
                  />
                  <ImageListItemBar
                    sx={{
                      background:
                        'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, ' +
                        'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
                    }}
                    title={data[39].name}
                    subtitle={`Total Comics: ${data[39].comics.available}`}
                    position="top"
                  />
                </Link>
              </ImageListItem>
            ) : null}
          </ImageList>

          <ImageList sx={{ width: 'auto', gap: '0 !important' }} variant="quilted" cols={2}>
            {loading
              ? data.map((v) => {
                  return (
                    <ImageListItem
                      sx={{ overflow: 'hidden' }}
                      key={v.id}
                      cols={v.cols || 1}
                      rows={v.rows || 1}
                    >
                      <Link to={`/${v.id}`}>
                        <img
                          src={`${v.thumbnail.path}.${v.thumbnail.extension}`}
                          alt={v.name}
                          loading="lazy"
                          style={{ width: '100%', height: '100%' }}
                        />
                        <ImageListItemBar
                          sx={{
                            background:
                              'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, ' +
                              'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
                          }}
                          title={v.name}
                          subtitle={`Total Comics: ${v.comics.available}`}
                          position="top"
                        />
                      </Link>
                    </ImageListItem>
                  );
                })
              : null}
          </ImageList>
        </div>
      ) : (
        <ImageList sx={{ width: 'auto', gap: '0 !important' }} variant="quilted" cols={2}>
          {loading
            ? data
                .filter((hero) => {
                  if (search === '') {
                    return hero;
                  } else if (hero.name.toLowerCase().includes(search.toLowerCase())) {
                    return hero;
                  }
                })
                .map((v) => {
                  return (
                    <ImageListItem
                      sx={{ overflow: 'hidden' }}
                      key={v.id}
                      cols={v.cols || 1}
                      rows={v.rows || 1}
                    >
                      <Link to={`/${v.id}`}>
                        <img
                          src={`${v.thumbnail.path}.${v.thumbnail.extension}`}
                          alt={v.name}
                          loading="lazy"
                          style={{ width: '100%', height: '100%' }}
                        />
                        <ImageListItemBar
                          sx={{
                            background:
                              'linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, ' +
                              'rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)',
                          }}
                          title={v.name}
                          subtitle={`Total Comics: ${v.comics.available}`}
                          position="top"
                        />
                      </Link>
                    </ImageListItem>
                  );
                })
            : null}
        </ImageList>
      )}
    </div>
  );
};

Home.propTypes = {
  search: PropTypes.string,
};

Home.defaultProps = {
  search: '',
};

export default Home;
