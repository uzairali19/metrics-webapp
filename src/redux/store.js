import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import marvelsReducer from './reducers/marvelReducer';

const reducer = combineReducers({
  marvelsReducer,
});

const store = createStore(reducer, compose(applyMiddleware(thunk, logger)));

export default store;
