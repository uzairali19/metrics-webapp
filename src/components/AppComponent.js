import React from 'react';
import { Routes, Route } from 'react-router-dom';
import Header from './Header';
import Home from './Home';
import Details from './Details';

const AppComponent = () => {
  const Hello = 'Hello World';
  return (
    <div>
      <Header />
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="details" element={<Details />} />
      </Routes>
    </div>
  );
};

export default AppComponent;
