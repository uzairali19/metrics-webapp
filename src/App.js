import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { getMarvels } from './redux/actions/marvels';
import AppComponent from './components/AppComponent';
import Loading from './components/Loading';

function App() {
  const [loading, setLoading] = useState(false);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getMarvels()).then(() => {
      setLoading(true);
    });
  }, []);

  return <div className="App">{loading ? <AppComponent /> : <Loading />}</div>;
}

export default App;
