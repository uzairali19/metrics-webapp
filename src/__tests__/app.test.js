import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import App from '../App';
import store from '../redux/store';
import Home from '../components/Home';
import AppComponent from '../components/AppComponent';
import Details from '../components/Details';
import Header from '../components/Header';

describe('Render App Component Correctly', () => {
  test('Hero App component', () => {
    const home = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(home).toMatchSnapshot();
  });
});

describe('Root Component render', () => {
  test('Marvel Home component', () => {
    const root = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App>
              <Home />
            </App>
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(root).toMatchSnapshot();
  });
});

describe('Main details Component render', () => {
  test('Marvel Hero Detail component', () => {
    const details = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App>
              <Details />
            </App>
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(details).toMatchSnapshot();
  });
});

describe('Header Component render', () => {
  test('Marvel Hero Header component', () => {
    const header = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App>
              <AppComponent>
                <Header />
                <Home />
                <Details />
              </AppComponent>
            </App>
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(header).toMatchSnapshot();
  });
});

describe('Search function render', () => {
  test('Render Search function for home and Header', () => {
    const header = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <App>
              <AppComponent>
                <Header />
                <Home search="IronMan" />
              </AppComponent>
            </App>
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(header).toMatchSnapshot();
  });
});
