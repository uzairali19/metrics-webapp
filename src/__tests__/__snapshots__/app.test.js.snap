// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component render Marvel Hero Header component 1`] = `
<div
  className="App"
>
  <div
    className="MuiBox-root css-ks2ip0"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1exiwx8-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
    <h2
      className="MuiTypography-root MuiTypography-h5 loading-text css-1vkkaf1-MuiTypography-root"
    >
      Loading ...
    </h2>
  </div>
</div>
`;

exports[`Main details Component render Marvel Hero Detail component 1`] = `
<div
  className="App"
>
  <div
    className="MuiBox-root css-ks2ip0"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1exiwx8-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
    <h2
      className="MuiTypography-root MuiTypography-h5 loading-text css-1vkkaf1-MuiTypography-root"
    >
      Loading ...
    </h2>
  </div>
</div>
`;

exports[`Render App Component Correctly Hero App component 1`] = `
<div
  className="App"
>
  <div
    className="MuiBox-root css-ks2ip0"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1exiwx8-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
    <h2
      className="MuiTypography-root MuiTypography-h5 loading-text css-1vkkaf1-MuiTypography-root"
    >
      Loading ...
    </h2>
  </div>
</div>
`;

exports[`Root Component render Marvel Home component 1`] = `
<div
  className="App"
>
  <div
    className="MuiBox-root css-ks2ip0"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1exiwx8-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
    <h2
      className="MuiTypography-root MuiTypography-h5 loading-text css-1vkkaf1-MuiTypography-root"
    >
      Loading ...
    </h2>
  </div>
</div>
`;

exports[`Search function render Render Search function for home and Header 1`] = `
<div
  className="App"
>
  <div
    className="MuiBox-root css-ks2ip0"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1exiwx8-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
    <h2
      className="MuiTypography-root MuiTypography-h5 loading-text css-1vkkaf1-MuiTypography-root"
    >
      Loading ...
    </h2>
  </div>
</div>
`;
